### ISSUE TITLE
Support for Typescript casting in `string` validator

### ISSUE THREAD

**Issue by victorgarciaesgi** [dwightjack/vue-types#134]
Hi and thanks for the 4.0 version! This is really time saving.

I found out that Typescript support has landed and made me use it! 
As i'm rewriting my component props (By the way, it's not very clear in the docs how to cast a type on a validator: `object<User>()`), I found out that `string()` does not support casting and that's a bummer because it is really usefull in a lot of cases!

Ex Before vue-types

```ts
export enum PopinAlignement {
  Center = 'Center',
  Start = 'Start',
  ContainerStart = 'ContainerStart',
  ContainerEnd = 'ContainerEnd',
  End = 'End',
  Full = 'Full',
}

export default defineComponent({
  props: {
     alignement: { type: String as PropType<PopinAlignement> },
  }
})
```

With vue-types

```ts
export default defineComponent({
  props: {
     // Error :(
     alignement: string<PopinAlignement>(),
  }
})
```

In vue-types, I can only type this as `String` and not an enum or a literal type like `type size = 'sm' | 'md'`

**Comment by dwightjack** [comment]
Hello, thanks for the feedback.

At the moment, generic arguments are supported in `array` and `object` validators only because I guessed they are the most common instances where you'd want to have TS-only checking to quickly set the prop without using validators like `shape`. 

For all other instances, using other validators like `oneOf` is the best choice because they provide runtime checks. So, for example you could have:

```ts
const size = oneOf(['sm', 'md'] as const) === 'sm' | 'md'
```

For enums, in TS 4.1+ you can use:

```ts
enum Hello {
  One = 'One',
  Two = 'Two',
}

const oneOfEnum = oneOf( Object.values<`${Hello}`>(Hello))
```

Or:

```ts
enum Hello {
  One = 'One',
  Two = 'Two',
}

const oneOfEnum = oneOf(Object.values<typeof Hello[keyof typeof Hello]>(Hello))
```

That said, I am also open to make the library more flexible if that improves the DX. I've tried a quick experiment, but I am getting some errors from the TS checker when I try to constraint the generic to `string` ([reference](https://github.com/dwightjack/vue-types/blob/ts-args-for-natives/src/validators/native.ts#L16-L19)).

Because the implementation is basically a function returning a `PropOptions` type, I've created a very basic sandbox that exposes the issue as well (i.e.: `stringProp` type check is failing) (https://codesandbox.io/s/tender-sammet-8k2lx?file=/src/index.ts). 

Because my knowledge of generics is weak, I'd be grateful if you feel to contribute to this feature 😃 .

**Comment by victorgarciaesgi** [comment]
Thanks for the great answer @dwightjack . I totally understand what you mean.
It's true that I want TS-only type checking because Volar(Vscode) will make the typecheck and it's a better DX overall.
I will still try using `oneOf`.

I'm working a solving the type issue on your sandbox, it's a very common and frustrating error with generics and subtypes that is not well documented by the Typescript team. I will keep you updated and I'm happy to contribute! 😄

**Comment by victorgarciaesgi** [comment]
@dwightjack One of the solution to bypass the TS error is to cast to unknown first. I made a fix to your sandbox using this solution!
https://codesandbox.io/s/stupefied-sky-7esxe?file=/src/index.ts

**Comment by dwightjack** [comment]
@victortotem thank you for your help! I've merged #135. I have opened another PR #136 adding the same generic argument to `number()` and `integer()`. 

I have also reviewed the documentation adding these new features and moving the `TYPESCRIPT HINT` info box just after the code example. Maybe in this way the TypeScript related documentation is a bit more prominent. 

If you want to check out the changes, I've published them as `vue-types@4.1.0-beta.1` (or `vue-types@next`)
